### ISSUE TITLE
Hercules-specific facility bits SNAFU

### ISSUE THREAD

**Issue by Fish-Git** [Fish-Git/hyperion#73]
Facility bits should be considered to be "set in stone".

Once a bit is defined as being at a particular location in the facilities list (bit number), it cannot (_must not!_) move. If e.g. bit 123 is defined as indicating facility foobar is installed, then it cannot (_must not!_) at some point in the future suddenly move to bit number 456 instead. Otherwise mainframe code previously written to expect it at the original location will malfunction.

&nbsp;
Unfortunately as originally coded by Jan, the "Hercules unique" facility bits were placed immediately following the highest defined IBM facility bit defined at that time (which at the time was bit 77; see header `esa390.h` in commit e0d05622a0c5f76f2a1af62c4a8d9d8e173939ce`(*))`.

This fatally flawed design was left unaddressed and additional Hercules-specific facility bits were continually added over time, until eventually new facility bits defined by IBM began overlapping the Hercules defined facility bits. (See page 4-100 of [SA22-7832-11 Principles of Operation](http://publibfp.dhe.ibm.com/epubs/pdf/dz9zr011.pdf)).

Commit aa80a17472ed73a6db6ea1d55db8a1bb851b30a5 was an attempt by yours truly to address this newly introduced Hercules vs. IBM facility bit collision/overlap, but unfortunately it simply moved the bits further down the bit-string (which as explained earlier is itself a huge no-no!), leaving the original design flaw unaddressed and leaving ourselves open to yet another collision/overlap problem at some point in the future. (The commit didn't fix/address the actual underlying problem but rather simply temporarily delayed it to some undetermined point in the future.)

&nbsp;
The bottom line is our Hercules-specific (Hercules-unique) facility bits need a permanent home that, once set, will from then on _**never**_ change. `(**)`

I'm thinking that since, as currently architected, the `STFLE` instruction supports a maximum of 16,384 facility bits (2K (2048) bytes), we might have no other choice but to move them to e.g. bit number 16256 (16384-128) and hope to hell IBM will never assign any facilities at, or beyond, that bit position.

&nbsp;
What do others think?
&nbsp;
___
`(*)`Although they were later moved (probably accidentally) to bit 128 instead, by commit c56110e8fb41d7c96616651558c7c593d6ecb6f8.

`(**)`As explained, facility bits _cannot_ move around! Once defined, they need to _stay where they're at!_ _**Forever!**_

**Comment by Fish-Git** [comment]
Unless anyone has a better idea I think I'm going to go with 16256.

**Comment by Fish-Git** [comment]
Fish wrote:

> Unless anyone has a better idea I think I'm going to go with 16256.

**_HOLD EVERYTHING!_**

I've just discovered an undocumented subtlety regarding the handling of Hercules-specific facility bits assigned to bit positions immediately following the highest supported (highest assigned) IBM facility bit.

In the only two places where the guest directly accesses facility bits (in the STFLE Store Facility List Extended instruction in esame.c and in the SIE Start Interpretive Execution instruction in sie.c) the code is written to prevent the guest from ever seeing or accessing the defined Hercules-specific faility bits, by virtue of _both_ them using the "number of IBM facility list bytes" value (`STFL_IBMBYSIZE`) and not the full/actual "number of facility list bytes actually defined by Hercules" value (`STFL_HERCBYSIZE`).

I completely missed that subtlety the when I first opened this issue.

Thus, even though the facility list array is defined/allocated in both SYSBLK and REGS using the full/complete "Herc" size:

* SYSBLK: `BYTE facility_list[GEN_MAXARCH][STFL_HERCBYSIZE];`
* REGS: `BYTE    facility_list[STFL_HERCBYSIZE];`

and the both archlvl.c `init_als()` and `set_alslevel()` functions always copy the full "Herc" length (`STFL_HERCBYSIZE`), it is because of the STFLE and SIE instructions always using `STFL_IBMBYSIZE` (the "IBM size") that prevents any IBM vs. Herc facility bit assignment conflict/overlap that I originally thought existed.

What I'm trying to say is this GitHub Issue "[Hercules-specific facility bits SNAFU](https://github.com/Fish-Git/hyperion/issues/73)" (i.e.  _this_ one that you're reading right now) is _**completely bogus!!**_  There is _no problem/conflict_ to be fixed!

I am going to close this issue and mark is as "invalid".

Sorry to waste anyone's time. &nbsp; `:(`
